sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,517 @@
+  // Represents any unrecognized character other than the above.
+  case class InvalidUnrecognizedCharacter(char: Char) extends InputToken
+}
+
+/**
+ * This class represents a parser to implement the to_number SQL function.
+ *
+ * It works by consuming an input string and a format string. This class accepts the format string
+ * as a field, and proceeds to iterate through the format string to generate a sequence of tokens
+ * (or throw an exception if the format string is invalid). Then when the function is called with an
+ * input string, this class steps through the sequence of tokens and compares them against the input
+ * string, returning a Scala Decimal object if they match (or throwing an exception otherwise).
+ *
+ * @param originNumberFormat the format string describing the expected inputs.
+ * @param errorOnFail true if evaluation should throw an exception if the input string fails to
+ *                    match the format string. Otherwise, returns NULL instead.
+ */
+class ToNumberParser(originNumberFormat: String, errorOnFail: Boolean) extends Serializable {
+  import ToNumberParser._
+
+  // Consumes the format string and produce a sequence of input tokens expected from each input
+  // string.
+  private lazy val inputTokens: Seq[InputToken] = {
+    var tokens = mutable.Seq.empty[InputToken]
+    var i = 0
+    var reachedDecimalPoint = false
+    val format = originNumberFormat
+    val len = originNumberFormat.length
+    while (i < len) {
+      val char: Char = originNumberFormat(i)
+      char match {
+        case ZERO_DIGIT =>
+          val prevI = i
+          do {
+            i += 1
+          } while (i < len && (format(i) == ZERO_DIGIT || format(i) == NINE_DIGIT))
+          if (reachedDecimalPoint) {
+            tokens :+= AtMostAsManyDigits(i - prevI)
+          } else {
+            tokens :+= ExactlyAsManyDigits(i - prevI)
+          }
+        case NINE_DIGIT =>
+          val prevI = i
+          do {
+            i += 1
+          } while (i < len && (format(i) == ZERO_DIGIT || format(i) == NINE_DIGIT))
+          tokens :+= AtMostAsManyDigits(i - prevI)
+        case POINT_SIGN | POINT_LETTER =>
+          tokens :+= DecimalPoint()
+          reachedDecimalPoint = true
+          i += 1
+        case COMMA_SIGN | COMMA_LETTER =>
+          tokens :+= ThousandsSeparator()
+          i += 1
+        case DOLLAR_LETTER | DOLLAR_SIGN =>
+          tokens :+= DollarSign()
+          i += 1
+        case OPTIONAL_PLUS_OR_MINUS_LETTER =>
+          tokens :+= OptionalPlusOrMinusSign()
+          i += 1
+        case OPTIONAL_MINUS_STRING_START if i < len - 1 &&
+          OPTIONAL_MINUS_STRING_END == originNumberFormat(i + 1) =>
+          tokens :+= OptionalMinusSign()
+          i += 2
+        case WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_START if i < len - 1 &&
+          WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_END == originNumberFormat(i + 1) =>
+          tokens +:= OpeningAngleBracket()
```

Comment:
is `+:=` prepend? maybe we should use explicit function names...

Reply:
Yes, but looking further, it turns out it returns a new collection every time, even though the `tokens` is an `mutable.Seq`! I changed it to a `mutable.Buffer` instead using explicit `append` and the code is simpler with guaranteed O(1) appends.